### ISSUE TITLE
DRAFT - Authentication (oAuth & Metamask)

### ISSUE THREAD

**Issue by srenault-meeds** [Meeds-io/MIPs#60]
## Rationale

When logging in the first time, steps are missing to help the user confirming its identity:
- For any authentication mode, it can be more than useful to request the user setting its profile avatar, display name (+ email in one case) in addition to the password
- When using metamask, confirming the email to be set in the profile must be requested. Currently, I can use a fake email. Thus, users won't get notifications and most importantly, in case we need to confirm its identity (to change the way he can logging in), we are not sure it is its email
- When using google or any other provider like OIDC, it could be more than interesting to collect these data (profile avatar and display name) to smooth the onboarding

In addition to this, once authenticated and using the platform, in case the user needs to update its authentication mode, it is not possible. This is unfortunate.

## 1. Functional Requirements

1. Administrate the authentication
* Admin access authentication settings to enable authentication mode
Mail & Metamask are set by default (for meeds package and only mail for other package)
Google and OIDC can be set using properties requested for these authentication (wellKnownConfigurationUrl, clientID, clientSecret, redirectURL)
Profile avatar, display name and email are collected / confirmed if possible when authenticating and admin can disable items except email.
> this needs to be extensible so we can collect more than these informations

2. End-users when logging in the first time
* Using metamask
Profile avatar, display name and email can be collected from ethereum name system
Users can confirm informations gathered
If not gathered, then users need to complete the informations needed
Once done, an email is sent to the added email so we can confirm the user identity (with a code or a token so he access the platform)

* Using mail
Profile avatar, display name can be set when logging in the first time
If invited from onboarding, he can confirm the informations
Once done, he access the platform without logging in again

* Using oAuth authentication (google, oidc)
Profile avatar, display name and email can be collected from the provider
Users can confirm informations gathered
If not gathered, then users need to complete the informations needed
Once done, he access the platform without logging in again

3. End-users can decide to use another authentication mode
Depending the provider proposed by admin, the user can access its personal settings to:
a. change its password (currently ok)
b. switch to another authentication provider
To complete this step, a confirmation needs to be sent to the user's email to confirm its identity with a code or with a token
Once confirmed, the user can request to use another provider
Once requested, a confirmation needs to be done:
* If metamask, signing needed and previous email remained
* If Google, confirmation sent to new account
* If OIDC, confirmation sent to new account (depending the mail used)

In addition to this, from the profile, the user can request to change its email.
To complete such thing, a confirmation needs to be sent to the current email to confirm the user's identity. Once done, another confirmation email will be sent to the new email to be used so we can confirm its identity.

### Feature Flags

| Property Name  | Default Value | Target Audience | Functional Behaviour |
| ------------- | ------------- | ------------- | ------------- |
|  |   |   |   |

### Access 
   - GUI 
   - API 

### Services & processing
   - Ex: Scheduled Jobs

### Data and persistence
   - Macro data model
   - Persistence layer 
   - Use of caches
## 5. Annexes
   - Reasoning that explain the technical choices
   - Impacts 
   - References

**Comment by srenault-meeds** [comment]
@boubaker @AzmiTouil if you want to take a look at it

**Comment by boubaker** [comment]
> When using metamask, confirming the email to be set in the profile must be requested. Currently, I can use a fake email. Thus, users won't get notifications and most importantly, in case we need to confirm its identity (to change the way he can logging in), we are not sure it is its email

I think it's already improved [here](https://builders.meeds.io/portal/meeds/tasks/taskDetail/1958)

**Comment by srenault-meeds** [comment]
> > When using metamask, confirming the email to be set in the profile must be requested. Currently, I can use a fake email. Thus, users won't get notifications and most importantly, in case we need to confirm its identity (to change the way he can logging in), we are not sure it is its email
> 
> I think it's already improved [here](https://builders.meeds.io/portal/meeds/tasks/taskDetail/1958)

Indeed. Only want to make sure the process is clear for our audience here and consistent with other use cases
